File: NaymNaymLevelUp/Rebuild/Child/TodayForestViewModel.swift
import CoreData
import Foundation

protocol TodayMealRepository: Sendable {
    func currentState(date: String) async -> MealLoadState
    func refresh(date: String, school: RebuildSchool) async
}

protocol TodayMealRecorder: Sendable {
    func execute(_ command: RecordMealCommand) async throws -> RecordMealResult
}

protocol TodayMealPhotoMetadataStore: Sendable {
    func photoIDs(
        date: String,
        normalizedMenuName: String
    ) async throws -> [String]
}

protocol TodayProgressProvider: Sendable {
    func totalXP() async throws -> Int
}

extension RebuildMealRepository: TodayMealRepository {}

struct TodayMealScheduleRepositoryAdapter: MealScheduleRepository {
    let repository: any TodayMealRepository

    func currentState(date: String) async -> MealLoadState {
        await repository.currentState(date: date)
    }

    func refresh(date: String, school: RebuildSchool) async {
        await repository.refresh(date: date, school: school)
    }
}

struct LiveTodayMealRecorder: TodayMealRecorder {
    let useCase: RecordMealUseCase

    func execute(_ command: RecordMealCommand) async throws -> RecordMealResult {
        try await Task.detached(priority: .userInitiated) {
            try useCase.execute(command)
        }.value
    }
}

struct EmptyTodayMealPhotoMetadataStore: TodayMealPhotoMetadataStore {
    func photoIDs(
        date: String,
        normalizedMenuName: String
    ) async throws -> [String] {
        []
    }
}

struct EmptyTodayProgressProvider: TodayProgressProvider {
    func totalXP() async throws -> Int {
        0
    }
}

actor CoreDataTodayProgressProvider: TodayProgressProvider {
    private let repository: RebuildProgressRepository

    init(container: NSPersistentContainer) {
        repository = RebuildProgressRepository(
            context: container.newBackgroundContext()
        )
    }

    func totalXP() async throws -> Int {
        let storedTotal = try repository.totalXP()
        guard let exactTotal = Int(exactly: storedTotal) else {
            throw RebuildRepositoryError.totalXPOverflow
        }
        return exactTotal
    }
}

struct CoreDataTodayMealPhotoMetadataStore: TodayMealPhotoMetadataStore {
    let container: NSPersistentContainer

    func photoIDs(
        date: String,
        normalizedMenuName: String
    ) async throws -> [String] {
        let context = container.newBackgroundContext()
        return try await context.perform {
            let request = NSFetchRequest<RebuildMealRecordManagedObject>(
                entityName: RebuildEntityName.mealRecord
            )
            request.predicate = NSCompoundPredicate(
                andPredicateWithSubpredicates: [
                    NSPredicate(format: "date == %@", date),
                    NSPredicate(
                        format: "normalizedMenuName == %@",
                        normalizedMenuName
                    ),
                ]
            )
            request.sortDescriptors = [
                NSSortDescriptor(key: "updatedAt", ascending: true),
                NSSortDescriptor(key: "id", ascending: true),
            ]

            var seen = Set<String>()
            var result: [String] = []
            for record in try context.fetch(request) {
                let values = try JSONDecoder().decode(
                    [String].self,
                    from: Data(record.photoIDsJSON.utf8)
                )
                for value in values where seen.insert(value).inserted {
                    result.append(value)
                }
            }
            return result
        }
    }
}

enum TodaySafetyAction: Equatable, Sendable {
    case allergyAvoided
    case guardianCheck
}

struct PreparedMealRecord: Equatable, Sendable {
    let command: RecordMealCommand
    let sourceMealDate: String
    let sourceItem: RebuildMealItem
    fileprivate let alternativeSelection: SameMealAlternativeSelection?

    var nutritionSnapshot: NutrientImpactSnapshot {
        guard let snapshot = command.nutritionSnapshot else {
            preconditionFailure("Prepared meal records require a nutrition snapshot.")
        }
        return snapshot
    }

    fileprivate init(
        command: RecordMealCommand,
        sourceMealDate: String,
        sourceItem: RebuildMealItem,
        alternativeSelection: SameMealAlternativeSelection?
    ) {
        precondition(command.nutritionSnapshot != nil)
        self.command = command
        self.sourceMealDate = sourceMealDate
        self.sourceItem = sourceItem
        self.alternativeSelection = alternativeSelection
    }
}

enum TodayForestError: Error, Equatable {
    case mealUnavailable
    case allergySafetyRequired
    case nutritionSnapshotUnavailable
    case stalePreparedRecord
}

@MainActor
final class TodayForestViewModel: ObservableObject {
    typealias Clock = @Sendable () -> Date

    static let activeStatuses: [RebuildEatingStatus] = [
        .finished,
        .half,
        .oneBite,
        .smelledOnly,
        .difficultToday,
        .allergyAvoided,
    ]

    static let difficultyReasonOrder: [RebuildDifficultyReason] = [
        .smell,
        .texture,
        .taste,
        .appearance,
        .other,
    ]

    let title = "오늘 급식"
    let primaryActionTitle = "오늘 급식 기록하기"

    @Published private(set) var meal: RebuildMealDay?
    @Published private(set) var sourceLabel = "급식을 확인하고 있어요"
    @Published private(set) var isPrimaryActionEnabled = false
    @Published private(set) var isLoading = false
    @Published private(set) var totalXP = 0
    @Published private(set) var lastGrantedXP = 0
    @Published private(set) var motion: RebuildMotionState = .idle
    @Published private(set) var motionRevision = 0
    @Published private(set) var message: String?
    @Published private(set) var lastNutritionGuidance: NutrientImpactGuidance?

    let dateText: String
    let dateKey: String
    let allergyCodes: [Int]

    private let repository: any TodayMealRepository
    private let recorder: any TodayMealRecorder
    private let photoMetadataStore: any TodayMealPhotoMetadataStore
    private let progressProvider: any TodayProgressProvider
    private let school: RebuildSchool?
    let isDemoMode: Bool
    private let now: Clock
    private let calendar: Calendar
    private var progressRevision = 0

    var mealScheduleRepository: any MealScheduleRepository {
        TodayMealScheduleRepositoryAdapter(repository: repository)
    }

    var detailSchool: RebuildSchool? {
        school
    }

    var isMealDetailActionEnabled: Bool {
        !isLoading
    }

    init(
        repository: any TodayMealRepository,
        recorder: any TodayMealRecorder,
        photoMetadataStore: any TodayMealPhotoMetadataStore =
            EmptyTodayMealPhotoMetadataStore(),
        progressProvider: any TodayProgressProvider =
            EmptyTodayProgressProvider(),
        school: RebuildSchool?,
        allergyCodes: [Int],
        isDemoMode: Bool = false,
        date: Date = Date(),
        calendar: Calendar = Calendar(identifier: .gregorian),
        now: @escaping Clock = { Date() }
    ) {
        self.repository = repository
        self.recorder = recorder
        self.photoMetadataStore = photoMetadataStore
        self.progressProvider = progressProvider
        self.school = school
        self.isDemoMode = isDemoMode
        self.allergyCodes = Array(Set(allergyCodes)).sorted()
        self.now = now

        var localizedCalendar = calendar
        if localizedCalendar.timeZone.identifier == "GMT" {
            localizedCalendar.timeZone =
                TimeZone(identifier: "Asia/Seoul") ?? .current
        }
        self.calendar = localizedCalendar
        let keyFormatter = DateFormatter()
        keyFormatter.calendar = localizedCalendar
        keyFormatter.locale = Locale(identifier: "en_US_POSIX")
        keyFormatter.timeZone = localizedCalendar.timeZone
        keyFormatter.dateFormat = "yyyy-MM-dd"
        dateKey = keyFormatter.string(from: date)

        let displayFormatter = DateFormatter()
        displayFormatter.calendar = localizedCalendar
        displayFormatter.locale = Locale(identifier: "ko_KR")
        displayFormatter.timeZone = localizedCalendar.timeZone
        displayFormatter.setLocalizedDateFormatFromTemplate("MMMMdEEEE")
        dateText = displayFormatter.string(from: date)
    }

    func recordingViewModel(
        for route: MealDayRoute
    ) -> TodayForestViewModel? {
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = calendar.timeZone
        formatter.dateFormat = "yyyy-MM-dd"
        guard let date = formatter.date(from: route.dateKey),
              formatter.string(from: date) == route.dateKey else {
            return nil
        }

        return TodayForestViewModel(
            repository: repository,
            recorder: recorder,
            photoMetadataStore: photoMetadataStore,
            progressProvider: progressProvider,
            school: school,
            allergyCodes: allergyCodes,
            isDemoMode: isDemoMode,
            date: date,
            calendar: calendar,
            now: now
        )
    }

    func load() async {
        isLoading = true
        let progressRevisionAtStart = progressRevision
        async let persistedTotalXP = try? progressProvider.totalXP()
        apply(await repository.currentState(date: dateKey))
        if let school {
            await repository.refresh(date: dateKey, school: school)
            apply(await repository.currentState(date: dateKey))
        }
        if let persistedTotalXP = await persistedTotalXP,
           progressRevision == progressRevisionAtStart {
            totalXP = persistedTotalXP
        }
        isLoading = false
    }

    func isAllergyRisk(_ item: RebuildMealItem) -> Bool {
        !Set(item.allergyCodes).isDisjoint(with: allergyCodes)
    }

    func synchronizeMeal(_ meal: RebuildMealDay?, for route: MealDayRoute) {
        guard route.dateKey == dateKey else { return }
        guard meal?.date == nil || meal?.date == dateKey else { return }
        self.meal = meal
        isPrimaryActionEnabled = meal?.menuItems.isEmpty == false
    }

    func isStatusEnabled(
        _ status: RebuildEatingStatus,
        for item: RebuildMealItem
    ) -> Bool {
        MealSafetyPolicy.allowedStatuses(
            childAllergyCodes: Set(allergyCodes),
            itemAllergyCodes: Set(item.allergyCodes)
        ).contains(status)
    }

    func prioritizedSafetyActions(
        for item: RebuildMealItem
    ) -> [TodaySafetyAction] {
        isAllergyRisk(item)
            ? [.allergyAvoided, .guardianCheck]
            : []
    }

    func prepareRecord(
        item: RebuildMealItem,
        status: RebuildEatingStatus,
        difficultyReasons: [RebuildDifficultyReason] = [],
        parentShareEnabled: Bool = false
    ) async throws -> PreparedMealRecord {
        guard let meal,
              meal.date == dateKey,
              meal.menuItems.filter({ $0 == item }).count == 1 else {
            throw TodayForestError.mealUnavailable
        }
        guard isStatusEnabled(status, for: item) else {
            throw TodayForestError.allergySafetyRequired
        }

        let normalizedMenuName = MealRecordIdentityNormalizer.normalizedMenuName(
            item.name
        )
        let photoIDs = try await photoMetadataStore.photoIDs(
            date: dateKey,
            normalizedMenuName: normalizedMenuName
        )
        let reasons = status == .difficultToday
            ? Self.orderedDifficultyReasons(difficultyReasons)
            : []
        let matchedAllergyCodes = Array(
            Set(item.allergyCodes).intersection(allergyCodes)
        ).sorted()
        let occurredAt = now()
        let recordID = "\(dateKey)|\(normalizedMenuName)"
        let visual = MealVisualResolver.resolve(item: item)
        let alternativeSelection: SameMealAlternativeSelection?
        if status == .difficultToday {
            alternativeSelection = SameMealAlternativeSelector.select(
                from: meal,
                currentItem: item,
                childAllergyCodes: allergyCodes
            )
        } else {
            alternativeSelection = nil
        }
        let nutrientIDs = Self.nutrientIDsForSnapshot(
            status: status,
            representativeNutrientIDs: visual.representativeNutrientIDs,
            alternativeTargetNutrientIDs:
                alternativeSelection?.provenance.targetNutrientIDs
        )
        let nutritionSnapshot: NutrientImpactSnapshot?
        if status == .difficultToday, let alternativeSelection {
            nutritionSnapshot = NutrientImpactSnapshotFactory.make(
                ruleVersion: visual.ruleVersion,
                recordID: recordID,
                date: dateKey,
                normalizedMenuName: normalizedMenuName,
                status: status,
                recordUpdatedAt: occurredAt,
                nutrientIDs: nutrientIDs,
                alternativeSelection: alternativeSelection
            )
        } else {
            nutritionSnapshot = NutrientImpactSnapshotFactory.make(
                ruleVersion: visual.ruleVersion,
                recordID: recordID,
                date: dateKey,
                normalizedMenuName: normalizedMenuName,
                status: status,
                recordUpdatedAt: occurredAt,
                nutrientIDs: nutrientIDs
            )
        }
        guard let nutritionSnapshot else {
            throw TodayForestError.nutritionSnapshotUnavailable
        }
        let command = RecordMealCommand(
            recordID: recordID,
            date: dateKey,
            menuName: item.name,
            status: status,
            difficultyReasons: reasons,
            allergyCodes: matchedAllergyCodes,
            childAllergyCodes: Array(Set(allergyCodes)).sorted(),
            itemAllergyCodes: Array(Set(item.allergyCodes)).sorted(),
            photoIDs: photoIDs,
            parentShareEnabled: parentShareEnabled,
            occurredAt: occurredAt,
            nutritionSnapshot: nutritionSnapshot
        )
        return PreparedMealRecord(
            command: command,
            sourceMealDate: meal.date,
            sourceItem: item,
            alternativeSelection: alternativeSelection
        )
    }

    static func nutrientIDsForSnapshot(
        status: RebuildEatingStatus,
        representativeNutrientIDs: [String],
        alternativeTargetNutrientIDs: [String]?
    ) -> [String] {
        guard status == .difficultToday else {
            return representativeNutrientIDs
        }
        return alternativeTargetNutrientIDs ?? representativeNutrientIDs
    }

    func record(
        prepared: PreparedMealRecord
    ) async throws -> RecordMealResult {
        try validateCurrentMeal(for: prepared)
        let result = try await recorder.execute(prepared.command)
        progressRevision += 1
        totalXP = result.totalXP
        lastGrantedXP = result.xpGranted
        motion = result.motion
        motionRevision += 1
        lastNutritionGuidance = result.nutritionGuidance
        message = result.xpGranted > 0
            ? "\(result.xpGranted) XP를 얻었어요!"
            : "오늘 기록을 저장했어요."
        return result
    }

    private func validateCurrentMeal(
        for prepared: PreparedMealRecord
    ) throws {
        let command = prepared.command
        let identity = MealRecordIdentityNormalizer.normalizedMenuName(
            prepared.sourceItem.name
        )
        guard let meal,
              meal.date == dateKey,
              prepared.sourceMealDate == dateKey,
              command.date == dateKey,
              command.recordID == "\(dateKey)|\(identity)" else {
            throw TodayForestError.stalePreparedRecord
        }
        let matchingItems = meal.menuItems.filter {
            MealRecordIdentityNormalizer.normalizedMenuName($0.name)
                == identity
        }
        guard matchingItems.count == 1,
              let currentItem = matchingItems.first else {
            throw TodayForestError.stalePreparedRecord
        }

        do {
            try MealSafetyPolicy.validate(
                childAllergyCodes: Set(allergyCodes),
                itemAllergyCodes: Set(currentItem.allergyCodes),
                status: command.status
            )
        } catch RecordMealError.allergySafetyRequired {
            throw TodayForestError.allergySafetyRequired
        }

        let currentChildAllergies = Array(Set(allergyCodes)).sorted()
        let currentItemAllergies = Array(Set(currentItem.allergyCodes)).sorted()
        let currentMatchedAllergies = Array(
            Set(currentChildAllergies).intersection(currentItemAllergies)
        ).sorted()
        let currentAlternativeSelection = command.status == .difficultToday
            ? SameMealAlternativeSelector.select(
                from: meal,
                currentItem: currentItem,
                childAllergyCodes: currentChildAllergies
            )
            : nil
        guard currentItem == prepared.sourceItem,
              command.childAllergyCodes == currentChildAllergies,
              command.itemAllergyCodes == currentItemAllergies,
              command.allergyCodes == currentMatchedAllergies,
              currentAlternativeSelection == prepared.alternativeSelection else {
            throw TodayForestError.stalePreparedRecord
        }
    }

    static func orderedDifficultyReasons(
        _ reasons: [RebuildDifficultyReason]
    ) -> [RebuildDifficultyReason] {
        let selected = Set(reasons)
        return difficultyReasonOrder.filter(selected.contains)
    }

    private func apply(_ state: MealLoadState) {
        switch state {
        case let .cached(cached, _):
            meal = cached
            sourceLabel = isDemoMode ? "체험 급식 · 저장됨" : "저장된 급식"
            message = "인터넷이 없어도 저장된 급식을 기록할 수 있어요."
        case let .refreshing(cached):
            meal = cached
            sourceLabel = cached == nil
                ? "급식을 확인하고 있어요"
                : (isDemoMode ? "체험 급식 · 업데이트 중" : "저장된 급식")
        case let .live(live):
            meal = live
            sourceLabel = isDemoMode ? "체험 급식" : "학교 급식"
            message = nil
        case .empty:
            meal = nil
            sourceLabel = "급식 정보 없음"
            message = "오늘 등록된 급식이 없어요."
        case let .failed(_, cached):
            meal = cached
            sourceLabel = cached == nil
                ? "급식을 불러오지 못했어요"
                : (isDemoMode ? "체험 급식 · 저장됨" : "저장된 급식")
            message = cached == nil
                ? "인터넷 연결을 확인하고 다시 시도해 주세요."
                : "인터넷이 없어 저장된 급식을 보여드려요."
        }
        isPrimaryActionEnabled = meal?.menuItems.isEmpty == false
    }
}

File: android/app/src/main/java/com/h19h29/naymnaymlevelup/rebuild/child/TodayForestViewModel.kt
package com.h19h29.naymnaymlevelup.rebuild.child

import com.fasterxml.jackson.core.JsonFactory
import com.h19h29.naymnaymlevelup.rebuild.data.RebuildDatabase
import com.h19h29.naymnaymlevelup.rebuild.meal.DifficultyReason
import com.h19h29.naymnaymlevelup.rebuild.meal.EatingStatus
import com.h19h29.naymnaymlevelup.rebuild.meal.MealDay
import com.h19h29.naymnaymlevelup.rebuild.meal.MealItem
import com.h19h29.naymnaymlevelup.rebuild.meal.MealLoadState
import com.h19h29.naymnaymlevelup.rebuild.meal.MealRepository
import com.h19h29.naymnaymlevelup.rebuild.meal.MealSafetyPolicy
import com.h19h29.naymnaymlevelup.rebuild.meal.MotionState
import com.h19h29.naymnaymlevelup.rebuild.meal.RecordMealCommand
import com.h19h29.naymnaymlevelup.rebuild.meal.RecordMealResult
import com.h19h29.naymnaymlevelup.rebuild.meal.RecordMealUseCase
import com.h19h29.naymnaymlevelup.rebuild.meal.School
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

interface TodayMealRepository {
    suspend fun currentState(date: String): MealLoadState
    suspend fun refresh(date: LocalDate, school: School)
}

class LiveTodayMealRepository(
    private val repository: MealRepository,
) : TodayMealRepository {
    override suspend fun currentState(date: String): MealLoadState =
        repository.currentState(date)

    override suspend fun refresh(date: LocalDate, school: School) {
        repository.refresh(date, school)
    }
}

fun interface TodayMealRecorder {
    suspend fun execute(command: RecordMealCommand): RecordMealResult
}

class LiveTodayMealRecorder(
    private val useCase: RecordMealUseCase,
) : TodayMealRecorder {
    override suspend fun execute(command: RecordMealCommand): RecordMealResult =
        useCase.execute(command)
}

fun interface TodayPhotoMetadataStore {
    suspend fun photoIDs(date: String, normalizedMenuName: String): List<String>
}

fun interface TodayProgressProvider {
    suspend fun totalXP(): Int
}

class RoomTodayPhotoMetadataStore(
    private val database: RebuildDatabase,
) : TodayPhotoMetadataStore {
    override suspend fun photoIDs(
        date: String,
        normalizedMenuName: String,
    ): List<String> {
        val seen = linkedSetOf<String>()
        database.mealRecordDao()
            .recordsForMenu(date, normalizedMenuName)
            .flatMap { StringArrayJson.decode(it.photoIdsJson) }
            .forEach(seen::add)
        return seen.toList()
    }
}

class RoomTodayProgressProvider(
    private val database: RebuildDatabase,
) : TodayProgressProvider {
    override suspend fun totalXP(): Int {
        val value = database.progressDao().totalXp()
        require(value in 0L..Int.MAX_VALUE.toLong()) {
            "Persisted XP is outside the supported range"
        }
        return value.toInt()
    }
}

enum class TodaySafetyAction {
    AllergyAvoided,
    GuardianCheck,
}

enum class TodayForestError {
    MealUnavailable,
    AllergySafetyRequired,
}

class TodayForestException(
    val reason: TodayForestError,
) : IllegalStateException(reason.name)

data class TodayForestUiState(
    val meal: MealDay? = null,
    val sourceLabel: String = "급식을 확인하고 있어요",
    val primaryActionEnabled: Boolean = false,
    val isLoading: Boolean = false,
    val totalXP: Int = 0,
    val lastGrantedXP: Int = 0,
    val motion: MotionState = MotionState.Idle,
    val motionRevision: Long = 0,
    val message: String? = null,
)

class TodayForestViewModel(
    private val repository: TodayMealRepository,
    private val recorder: TodayMealRecorder,
    private val photoMetadataStore: TodayPhotoMetadataStore =
        TodayPhotoMetadataStore { _, _ -> emptyList() },
    private val progressProvider: TodayProgressProvider =
        TodayProgressProvider { 0 },
    private val school: School?,
    allergyCodes: List<Int>,
    val date: LocalDate = LocalDate.now(ZoneId.of("Asia/Seoul")),
    private val now: () -> Instant = Instant::now,
) {
    val title = "오늘 급식"
    val primaryActionTitle = "오늘 급식 기록하기"
    val dateKey: String = date.toString()
    val dateText: String = date.format(
        DateTimeFormatter.ofPattern("M월 d일 EEEE", Locale.KOREAN),
    )
    val allergyCodes: List<Int> = allergyCodes.distinct().sorted()

    private val mutableState = MutableStateFlow(TodayForestUiState())
    val state: StateFlow<TodayForestUiState> = mutableState.asStateFlow()
    private var progressRevision = 0L

    suspend fun load() = coroutineScope {
        mutableState.value = mutableState.value.copy(isLoading = true)
        val revisionAtStart = progressRevision
        val persistedTotal = async {
            runCatching { progressProvider.totalXP() }.getOrNull()
        }
        apply(repository.currentState(dateKey))
        school?.let {
            repository.refresh(date, it)
            apply(repository.currentState(dateKey))
        }
        persistedTotal.await()?.let { total ->
            if (progressRevision == revisionAtStart) {
                mutableState.value = mutableState.value.copy(totalXP = total)
            }
        }
        mutableState.value = mutableState.value.copy(isLoading = false)
    }

    fun isAllergyRisk(item: MealItem): Boolean =
        item.allergyCodes.any(allergyCodes::contains)

    fun isStatusEnabled(status: EatingStatus, item: MealItem): Boolean =
        status in MealSafetyPolicy.allowedStatuses(
            childAllergyCodes = allergyCodes.toSet(),
            itemAllergyCodes = item.allergyCodes.toSet(),
        )

    fun prioritizedSafetyActions(item: MealItem): List<TodaySafetyAction> =
        if (isAllergyRisk(item)) {
            listOf(
                TodaySafetyAction.AllergyAvoided,
                TodaySafetyAction.GuardianCheck,
            )
        } else {
            emptyList()
        }

    suspend fun record(
        item: MealItem,
        status: EatingStatus,
        difficultyReasons: List<DifficultyReason> = emptyList(),
        parentShareEnabled: Boolean = false,
    ): RecordMealResult {
        if (mutableState.value.meal == null) {
            throw TodayForestException(TodayForestError.MealUnavailable)
        }
        if (!isStatusEnabled(status, item)) {
            throw TodayForestException(TodayForestError.AllergySafetyRequired)
        }
        val normalizedName = item.name.trim().lowercase(Locale.ROOT)
        val photoIDs = photoMetadataStore.photoIDs(dateKey, normalizedName)
        val reasons = if (status == EatingStatus.DifficultToday) {
            orderedDifficultyReasons(difficultyReasons)
        } else {
            emptyList()
        }
        val childAllergyCodes = allergyCodes.toSet().sorted()
        val itemAllergyCodes = item.allergyCodes.toSet().sorted()
        val matchedAllergies = childAllergyCodes
            .intersect(itemAllergyCodes.toSet())
            .sorted()
        val result = recorder.execute(
            RecordMealCommand(
                recordID = "$dateKey|$normalizedName",
                date = dateKey,
                menuName = item.name,
                status = status,
                difficultyReasons = reasons,
                allergyCodes = matchedAllergies,
                childAllergyCodes = childAllergyCodes,
                itemAllergyCodes = itemAllergyCodes,
                photoIDs = photoIDs,
                parentShareEnabled = parentShareEnabled,
                occurredAt = now(),
            ),
        )
        progressRevision += 1
        mutableState.value = mutableState.value.copy(
            totalXP = result.totalXP,
            lastGrantedXP = result.xpGranted,
            motion = result.motion,
            motionRevision = progressRevision,
            message = if (result.xpGranted > 0) {
                "${result.xpGranted} XP를 얻었어요!"
            } else {
                "오늘 기록을 저장했어요."
            },
        )
        return result
    }

    private fun apply(mealState: MealLoadState) {
        mutableState.value = when (mealState) {
            is MealLoadState.Cached -> mutableState.value.copy(
                meal = mealState.meal,
                sourceLabel = "저장된 급식",
                primaryActionEnabled = mealState.meal.menuItems.isNotEmpty(),
                message = "인터넷이 없어도 저장된 급식을 기록할 수 있어요.",
            )
            is MealLoadState.Refreshing -> mutableState.value.copy(
                meal = mealState.cached,
                sourceLabel = if (mealState.cached == null) {
                    "급식을 확인하고 있어요"
                } else {
                    "저장된 급식"
                },
                primaryActionEnabled = mealState.cached?.menuItems?.isNotEmpty() == true,
            )
            is MealLoadState.Live -> mutableState.value.copy(
                meal = mealState.meal,
                sourceLabel = "학교 급식",
                primaryActionEnabled = mealState.meal.menuItems.isNotEmpty(),
                message = null,
            )
            MealLoadState.Empty -> mutableState.value.copy(
                meal = null,
                sourceLabel = "급식 정보 없음",
                primaryActionEnabled = false,
                message = "오늘 등록된 급식이 없어요.",
            )
            is MealLoadState.Failed -> mutableState.value.copy(
                meal = mealState.cached,
                sourceLabel = if (mealState.cached == null) {
                    "급식을 불러오지 못했어요"
                } else {
                    "저장된 급식"
                },
                primaryActionEnabled = mealState.cached?.menuItems?.isNotEmpty() == true,
                message = if (mealState.cached == null) {
                    "인터넷 연결을 확인하고 다시 시도해 주세요."
                } else {
                    "인터넷이 없어 저장된 급식을 보여드려요."
                },
            )
        }
    }

    companion object {
        val activeStatuses = listOf(
            EatingStatus.Finished,
            EatingStatus.Half,
            EatingStatus.OneBite,
            EatingStatus.SmelledOnly,
            EatingStatus.DifficultToday,
            EatingStatus.AllergyAvoided,
        )
        val difficultyReasonOrder = listOf(
            DifficultyReason.Smell,
            DifficultyReason.Texture,
            DifficultyReason.Taste,
            DifficultyReason.Appearance,
            DifficultyReason.Other,
        )

        fun orderedDifficultyReasons(
            reasons: List<DifficultyReason>,
        ): List<DifficultyReason> {
            val selected = reasons.toSet()
            return difficultyReasonOrder.filter(selected::contains)
        }
    }
}

private object StringArrayJson {
    private val factory = JsonFactory()

    fun decode(raw: String): List<String> =
        factory.createParser(raw).use { parser ->
            val result = mutableListOf<String>()
            if (parser.nextToken() == com.fasterxml.jackson.core.JsonToken.START_ARRAY) {
                while (
                    parser.nextToken() !=
                    com.fasterxml.jackson.core.JsonToken.END_ARRAY
                ) {
                    parser.valueAsString?.let(result::add)
                }
            }
            result
        }
}
